### PR TITLE
Optimize specialization phase

### PR DESCRIPTION
**Pull request by retronym**

### CODE REVIEW

**Comment by retronym** [comment]
I'm seeing good benchmark results from this in https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/2473/console, so pushing my WIP to commit myself to cleaning it up enough for merging.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -72,8 +74,47 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
   override def changesBaseClasses = true
   override def keepsTypeParams = true
 
-  type TypeEnv = immutable.Map[Symbol, Type]
-  def emptyEnv: TypeEnv = Map[Symbol, Type]()
+  final case class TypeEnvWrapper(value: immutable.Map[Symbol, Type]) {
```
Comment (diesalbla) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:77]: Can this be turned into an `AnyVal`?

Comment (retronym) [reply src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:77]: Nope:

  - it's not a top level class (although that could be solved with some work by de-caking it)
  - it has a field to cache the unzipped lists

**Comment by retronym** [comment]
Benchmark [results](https://scala-ci.typesafe.com/grafana/dashboard/db/scala-benchmark?orgId=1&from=1477761925988&to=1552540112105&var-branch=All&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40).

The pronounced benefits are from the tiny compilation job, which still trigger a lot of specialization info transforms on the collections hierarchy.

I tracked down one of test failures in this PR. It was arguably just a problem in the test, which was looking at class infos after a compilation run to make assertions about parameter field accessors reusing fields in superclasses. The second commit in this PR avoids creating a new entry in the type history when specialization is found to be a no-op, and [subsequent mutation of the decls ](https://github.com/scala/scala/blob/ed41f9659ced08c2e1b64021ab22a3b04107a279/src/compiler/scala/tools/nsc/transform/Constructors.scala#L779-L782) in `constructors`, which is now visible even in `exitingTyper(cls.info)`. `constructors` itself should probably have been written as an info-transform phase. 

I haven't tracked down the other failure.

**Comment by adriaanm** [comment]
Perhaps too ambitious for 2.12 in its late stage in the maintenance cycle?

**Comment by retronym** [comment]
@adriaanm Yep, this needs some more work or retargetting to satisfy the risk/reward. I'll shuffle it back to my TODO list for now.

**Pull request closed without merging**